elastic/apm-server
# stream/processor: Limit concurrent decoders to 200

## Motivation/summary

Adds a new configurable (for now) setting which allows us to limit the
maximum number of requests that are being decoded at any time. This
effectively limits the maximum memory consumption of the intake layer
of the server to `300KB * 10 events * 200 concurrent = 586MB`, of course
the actual maximum heap representation may be more, so adding a 20% of
overhead in this calculation still yields the ~704MB of memory usage for
the decoders, which is the maximum they will ever use. This plus a ~50MB
for the caches of all the bulk indexer (5MB * 10 bulk indexer) is still
below the 1GB mark. Even if Fleet and Elastic Agent were to use 100MB
we're still ~860 MB of memory used, which is below the smallest cgroup
memory limit size in ESS.

I've done some preliminary benchmarking and profiling on this change and
it seems to improve the lock contention on the `indexer.activeMu` since
the number of goroutines trying to acquire or release the lock is lower.

The indexing throughput didn't seem to be affected and the ingested
events/s throughput seems to be the same as it was before, except the
memory consumption is much more contained, only having observed ~750MB
of total RSS consumption. `8.1.2` went all the way up to ~1100MB RSS and
seemed to have slightly lower throughput (most likely due to the extra
contention that could be explained by a higher number of goroutines
trying to acquire the indexer.activeMu).

As it can be observed below, benchmarked with 550 agents and 1200, the
performance was much worse on `8.1.2` and the memory consumption also
much higher (1800-2200MB).

```
BenchmarkAgentGo-550     7346 events/sec  1827118 B/op  5236 allocs/op
BenchmarkAgentNodeJS-550 5452 events/sec  2605381 B/op  4225 allocs/op
BenchmarkAgentPython-550 5529 events/sec  9050536 B/op  22003 allocs/op
BenchmarkAgentRuby-550   6528 events/sec  2887096 B/op	6116 allocs/op
```

Tested with the APM Server standalone running on a `c5d.large` instance,
and a backing 2 node 8GB Elasticsearch cluster running in ESS in the
same region.

## Checklist


## How to test these changes

* Run `apmbench` with a medium to high number of `-agents` 300-800 agents should be enough.
* Verify that the `apm-server`'s memory doesn't exceed 800MB.

## Related issues

Closes #7795 

## Profiles

### Goroutine

#### 8.1.2

![8.1.2-goroutine](https://user-images.githubusercontent.com/7286993/162067393-042a901b-277b-43cf-ab4a-cb42dadb8a71.png)

### Patch

![patch-goroutine](https://user-images.githubusercontent.com/7286993/162067862-4119b711-2fff-4c65-b97d-6dda0497fbcf.png)

### Goroutine

#### 8.1.2

![8.1.2-heap](https://user-images.githubusercontent.com/7286993/162068061-5e1bc131-0f82-40ea-9791-78fa6cc5f552.png)

### Patch

![patch-heap](https://user-images.githubusercontent.com/7286993/162069224-e6576c45-f32b-46d2-94f6-028f5bf90a51.png)
